evm/src/trace/identifier/etherscan.rs

```diff
@@ -0,0 +1,177 @@
+    pub fn new(
+        chain: Option<impl Into<Chain>>,
+        etherscan_api_key: Option<String>,
+        cache_path: Option<PathBuf>,
+        ttl: Duration,
+    ) -> Self {
+        if let Some(cache_path) = &cache_path {
+            if let Err(err) = std::fs::create_dir_all(cache_path.join("sources")) {
+                warn!(target: "etherscanidentifier", "could not create etherscan cache dir: {:?}", err);
+            }
+        }
+
+        Self {
+            client: chain.and_then(|chain| {
+                etherscan_api_key.and_then(|key| {
+                    etherscan::Client::new_cached(chain.into(), key, cache_path, ttl).ok()
+                })
+            }),
+        }
+    }
+}
+
+impl TraceIdentifier for EtherscanIdentifier {
+    fn identify_addresses(
+        &self,
+        addresses: Vec<(&Address, Option<&Vec<u8>>)>,
+    ) -> Vec<AddressIdentity> {
+        self.client.as_ref().map_or(Default::default(), |client| {
+            let mut fetcher = EtherscanFetcher::new(client.clone(), Duration::from_secs(1), 5);
+
+            for (addr, _) in addresses {
+                fetcher.push(*addr);
+            }
+
+            let fut = fetcher
+                .map(|(address, label, abi)| AddressIdentity {
+                    address,
+                    label: Some(label.clone()),
+                    contract: Some(label),
+                    abi: Some(Cow::Owned(abi)),
+                })
+                .collect();
+
+            foundry_utils::RuntimeOrHandle::new().block_on(fut)
+        })
+    }
+}
+
+type EtherscanFuture =
+    Pin<Box<dyn Future<Output = (Address, Result<ContractMetadata, EtherscanError>)>>>;
+
+/// A rate limit aware Etherscan client.
+///
+/// Fetches information about multiple addresses concurrently, while respecting rate limits.
+pub struct EtherscanFetcher {
+    /// The Etherscan client
+    client: etherscan::Client,
+    /// The time we wait if we hit the rate limit
+    timeout: Duration,
+    /// The interval we are currently waiting for before making a new request
+    backoff: Option<Interval>,
+    /// The maximum amount of requests to send concurrently
+    concurrency: usize,
+    /// The addresses we have yet to make requests for
+    queue: Vec<Address>,
+    /// The in progress requests
+    in_progress: FuturesUnordered<EtherscanFuture>,
+}
+
+impl EtherscanFetcher {
+    pub fn new(client: etherscan::Client, timeout: Duration, concurrency: usize) -> Self {
+        Self {
+            client,
+            timeout,
+            backoff: None,
+            concurrency,
+            queue: Vec::new(),
+            in_progress: FuturesUnordered::new(),
+        }
+    }
+
+    pub fn push(&mut self, address: Address) {
+        self.queue.push(address);
+    }
+
+    fn queue_next_reqs(&mut self) {
+        if self.in_progress.len() < self.concurrency {
+            while let Some(addr) = self.queue.pop() {
```

Comment:
no because the queue might be bigger than the amount of requests we want to have in flight, so we should break as soon as we hit that number

when a future is pushed to `FuturesUnordered` and we then call `poll` it polls all the futures